blockchain/chain_sync/src/sync.rs

```diff
@@ -461,13 +463,54 @@ where
         }
         Ok(())
     }
+    /// fork detected, collect tipsets to be included in return_set sync_headers_reverse
+    async fn sync_fork(&mut self, head: &Tipset, to: &Tipset) -> Result<Vec<Tipset>, Error> {
+        // TODO change from using random peerID to managed
+        let peer_id = PeerId::random();
+        // pulled from Lotus: https://github.com/filecoin-project/lotus/blob/master/chain/sync.go#L996
+        const FORK_LENGTH_THRESHOLD: u64 = 500;
+
+        // Load blocks from network using blocksync
+        let tips: Vec<Tipset> = self
+            .network
+            .blocksync_headers(peer_id.clone(), head.parents(), FORK_LENGTH_THRESHOLD)
+            .await
+            .map_err(|_| Error::Other("Could not retrieve tipset".to_string()))?;
+
+        let mut ts = self.chain_store.tipset_from_keys(to.parents())?;
+
+        for i in 0..tips.len() {
+            if *ts.epoch().to_u64() == 0 {
+                if self.genesis.blocks() != ts.blocks() {
+                    return Err(Error::Other(
+                        "Chain is linked back to a different genesis (bad genesis)".to_string(),
+                    ));
+                }
+                return Err(Error::Other(
+                    "Synced chain forked at genesis, refusing to sync".to_string(),
+                ));
+            }
+            if ts == tips[i] {
+                return Ok(tips[0..=i].to_vec());
+            }
+        }
+
+        let mut n = 1;
+
+        while n < tips.len() {
+            if ts.epoch() > tips[n].epoch() {
+                ts = self.chain_store.tipset_from_keys(ts.parents())?;
+            } else {
+                n += 1;
+            }
+        }
 
-    async fn sync_fork(&mut self, _head: &Tipset, _to: &Tipset) -> Result<Vec<Arc<Tipset>>, Error> {
-        // TODO sync fork until tipsets are equal or reaches genesis
-        todo!()
+        Err(Error::Other(
+            "Fork longer than threshold finality of 500".to_string(),
+        ))
```

Comment:
I'm so confused on what this is doing, you are returning Error here regardless, why are you doing this loop? What, at a high level, are you trying to do here?

Reply:
updated